lib/portlayer/event/manager.go

```diff
@@ -131,16 +131,19 @@ func (mgr *Manager) Subscribed() int {
 
 // Publish events to subscribers
 func (mgr *Manager) Publish(e events.Event) {
-
-	// subscribers for this event
-	subs := mgr.subs.subscribers[e.Topic()]
-
-	log.Debugf("Found %d subscribers to %s", len(subs), e.Topic())
-
-	for sub, f := range subs {
-		//TODO: need to introduce timeout -- either with callbacks
-		// or when channels are introduced
-		log.Debugf("Event manager calling back to %s", sub)
-		f(e)
-	}
+	// TODO: this will not block, but might still want to consider
+	// a timeout for the callback
+	go func() {
+		mgr.subs.mu.RLock()
+		defer mgr.subs.mu.RUnlock()
```

Comment:
You actually don't need to hold the lock for the whole time the subscribers are notified; you can release it right after you get `subs`.